ttnn/cpp/ttnn/operations/moreh/moreh_mean/device/moreh_mean_device_operation.cpp

```diff
@@ -0,0 +1,128 @@
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    validate_tensors(operation_attributes, tensor_args);
+};
+
+MorehMeanOperation::shape_return_value_t MorehMeanOperation::compute_output_shapes(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto input_shape = tensor_args.input.get_shape();
+    auto output_shape = input_shape;
+    auto input_rank = input_shape.rank();
+
+    auto dim = operation_attributes.dim;
+
+    if (operation_attributes.keepdim) {
+        auto padding = output_shape.value.padding();
+        if (dim + 1 == input_rank) {
+            output_shape.value[dim] = tt::constants::TILE_WIDTH;
+            padding[dim] = Padding::PadDimension{0, 31};
+        } else if (dim + 2 == input_rank) {
+            output_shape.value[dim] = tt::constants::TILE_HEIGHT;
+            padding[dim] = Padding::PadDimension{0, 31};
+        } else {
+            output_shape.value[dim] = 1;
+        }
+
+        return Shape(tt::tt_metal::Shape(output_shape.value, padding));
+    }
+
+    std::vector<uint32_t> shape;
+    std::vector<Padding::PadDimension> pad_dimensions;
+    const bool is_tile_dim = (dim == input_rank - 1 || dim == input_rank - 2);
+    const std::size_t output_rank = (is_tile_dim) ? (input_rank) : (input_rank - 1);
+    auto input_padding = input_shape.value.padding();
+
+    // e.g. (2, 64, 64) with dim 1 to be (2, 1[32], 64)
+    // e.g. (2, 64, 64) with dim 0 to be (64, 64)
+    for (int i = 0; i < input_rank; ++i) {
+        bool is_reduced_dim = (i == dim);
+        if (is_reduced_dim && !is_tile_dim)
+            continue;
+
+        shape.push_back((is_reduced_dim && is_tile_dim) ? (tt::constants::TILE_HEIGHT) : (input_shape.value[i]));
+        pad_dimensions.push_back((is_reduced_dim && is_tile_dim) ? (Padding::PadDimension{0, 31}) : (input_padding[i]));
+    }
+
+    auto padding = Padding(pad_dimensions, input_padding.pad_value());
+    return Shape(tt::tt_metal::Shape(shape, padding));
+}
+
+MorehMeanOperation::tensor_return_value_t MorehMeanOperation::create_output_tensors(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto& output = tensor_args.output;
+    if (output.has_value()) {
+        return {output.value()};
+    }
+
+    return create_device_tensor(
+        compute_output_shapes(operation_attributes, tensor_args),
+        tensor_args.input.get_dtype(),
+        tensor_args.input.get_layout(),
+        tensor_args.input.device(),
+        operation_attributes.output_mem_config);
+}
+
+std::tuple<MorehMeanOperation::operation_attributes_t, MorehMeanOperation::tensor_args_t> MorehMeanOperation::invoke(
+    const Tensor& input,
+    const int64_t dim,
+    const bool keepdim,
+    const std::optional<uint32_t>& divisor,
+    const std::optional<Tensor>& output,
+    const std::optional<MemoryConfig>& output_mem_config,
```

Comment:
```suggestion
    const std::optional<MemoryConfig>& output_memory_config,
```
It seems that TT prefers `memory_config` over `mem_config`.